docs/proposals/202005_query_logging.md

```diff
@@ -0,0 +1,242 @@
+
+![](../img/thanos_log_limit.png)
+
+Each of the use case of the sub-part would be described and a possible implementation would be discussed -
+
+#### Using Middlewares -
+
+Since Middlewares are a good chunk of the implementation, it makes sense to discuss briefly about -
+
+* How we are going to use the middlewares in our logging.
+* Discuss the different types of middlewares that we plan to use.
+
+**Usage of Middlewares** : As it has been suggested by the diagram above, we plan to put in interceptors/middlewares to intercept the connection for a server implementation and client implementation.
+
+The middlewares helps in intercepting the queries flowing before a client, and all the queries leaving the server.
+
+The middlewares provide a nice abstraction to log the queries because we have the following objective from the logging -
+
+* Log all the queries
+* Log some specific queries
+* Logging at a certain level(Debug / Info)
+* Have a same request-id for a given request
+
+All of the goals are easily possible using grpc middlewares, as the grpc-ecosystem has a nice set of middlewares that makes the above possible, so we can just plug and play our middlewares
+
+For HTTP, all these nice middlewares are not present, so we need to code up our implementation performing the same logic.
+
+**Types of Middlewares** : We are going to use the following middlewares for achieving the above objectives -
+
+* [`logging.WithDecider`](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v2/interceptors/logging/options.go#L51) - This `grpc.Option` would help us in deciding the logic whether a given request should be logged or not. This should help in logging specific / all queries
+* [`logging.WithLevels`](https://github.com/grpc-ecosystem/go-grpc-middleware/blob/v2/interceptors/logging/options.go#L58) - This `grpc.Option` would help us in fixating the level of query that we might want to log. So based on a query hitting a certain criteria, we can switch the levels of it
+* For using the request-id, we can store the request-id in the metadata of the `context`, and while logging, we can use it.
+* Same equivalent for the HTTP can be implemented for mirroring the logic of gRPC middlewares
+
+#### Use case 1: Audit Logging
+
+**Audit logging** is a kind of logging, where we log in every internal API requests. Typically, rather than logging in specific queries, we ought to log in every query that has been made until now.
+
+From a developer’s perspective, audit logs can keep you sane by giving you some insight into how a complex system arrived at its current state. Developers are typically fond of state diagrams and deterministic outcomes. An audit log that shows that a person or process applied a change can give developers valuable insight.
+
+#### Implementation:
+
+
+```go
+// The request-id would be used from the context of the request
+auditLogger = log.With(logger, "component", component.StoreAPI)
```

Comment:
Is audit logger logged at the end of request or start? or both?

Reply:
My concern here in audit logging, is that it feels like this is just another level of adaptive logging.

So you generally have request logging. and options:
1. No request logging
1. Adaptive on errors only
1. Adaptive on errors plus slow requests
1. Adaptive if you switch quick runtime flag
1. ...
1. Log all = audit.

Reply:
Then we only have a request logger (or many) targeting certain file/output, with different adaptation or audit (full log)

Reply:
Done